### ISSUE TITLE
Consider with-members returning same instance when unmodified

### ISSUE THREAD

**Issue by atifaziz** [amis92/RecordGenerator#63]
When I have implemented records types by-hand, the `With*` member have been coded to return the same instance if the new value is the same as current and it's a lot of boilerplate code. Following is an illustrated example:

```c#
sealed class Record
{
    public int    Integer { get; }
    public string String  { get; }
    
    public Record(int integer, string @string) =>
        (Integer, String) = (integer, @string);
        
    public Record WithInteger(int value) =>
        Integer == value ? this : new Record(value, String);

    public Record WithString(string value) =>
        String == value ? this : new Record(Integer, value);
}
```

This is especially useful for record types representing options. For a real example, see [`AwaitQueryOptions` from my MoreLINQ project](https://github.com/morelinq/MoreLINQ/blob/5205ea241d72b079436060d330cd5c2eae7cdcdf/MoreLinq/Experimental/Await.cs#L32-L113).

I would like to see RecordGenerator help with such no-brainer boilerplate. It can go a long way to avoid over-allocating at the cost of fair and shallow equality tests (most of the time). It can make immutability cheaper.

I have not thought about how the design of this would look like RecordGenerator; that is, would it be merely another feature flag on `[Record]` and/or offer property-level attributes allowing finer control.

**Comment by amis92** [comment]
It is a nice optimization I've been thinking about at some point as well. I think it should be the new behavior (non-optional). I can't imagine a case where one would depend on old behavior (especially since it'd be the same as MemberwiseClone).

However if you feel like this should be behind a switch, I'd ask for this to be included in the most basic feature set as well (in terms of flags).

**Comment by atifaziz** [comment]
> However if you feel like this should be behind a switch, I'd ask for this to be included in the most basic feature set as well (in terms of flags).

I agree it should be the default behavior. A flag, if ever needed, could be added in a future version.
